### PR TITLE
Feature/signIn Code  Review

### PR DESCRIPTION
**Pull request by PrahladSingh-coditas**
### This pull request completes the feature/signIn functionality by implementing dependency injection for better code modularity and testability.

**Changes Made:**
- Implemented dependency injection in the signIn feature to decouple components and promote easier unit testing.
- Done with few minor changes as suggested by mentors.

Screenshot 2024-04-10 at 5 47 57 PM

### CODE REVIEW

File: src/app/authentication/constants/constants.go [hunk 1]
```diff
@@ -1,13 +1,7 @@
...
+)
 
-// var UserInstance = models.UserInfo{}
-
-// dbHost := viper.GetString("database.host")
-// dbPort := viper.GetInt("database.port")
-// dbUsername := viper.GetString("database.username")
-// dbPassword := viper.GetString("database.password")
-// dbName := viper.GetString("database.name")
+var UserInstance = models.UserInfo{}
```
Comment (vipin-raj-coditas) [inline src/app/authentication/constants/constants.go]: Don't define UserInstance here.

File: src/app/authentication/handlers/signUpHandler.go [hunk 2]
```diff
@@ -1,45 +1,36 @@
...
-		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid email format"})
+	// Call the SignUp method to handle user signup
+	if err := userService.SignUp(&user); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
 		return
 	}
-	log.Println("Email validated!")
+
+	// Respond with success message
+	c.JSON(http.StatusOK, gin.H{"message": "User information inserted successfully"})
```
Comment (vipin-raj-coditas) [inline src/app/authentication/handlers/signUpHandler.go:36]: Store all error messages in error.go and all constants in constants.go

File: src/app/authentication/models/signInModel.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package models
+
+type SignInCredentials struct {
+	UserName string `json:"name" validate:"required"`
+	Password string `json:"password" validate:"required"`
```
Comment (vipin-raj-coditas) [inline src/app/authentication/models/signInModel.go]: Add all possible validation for password

File: src/app/authentication/models/signUpModel.go [hunk 4]
```diff
@@ -4,11 +4,19 @@ type UserInfo struct {
 	ID            int32  `json:"user_id"`
 	Name          string `json:"user_name"`
 	Email         string `json:"user_email"`
-	PhoneNumber   int32  `json:"user_phoneNumber"`
-	PanCardNumber int32  `json:"user_panCardNumber"`
+	PhoneNumber   string `json:"user_phoneNumber"` // Changed type to string
+	PanCardNumber string `json:"user_panCardNumber"`
 	Password      string `json:"user_password"`
```
Comment (vipin-raj-coditas) [inline src/app/authentication/models/signUpModel.go]: Keep all json name as camelCase as best practice and also add all possible in built and custom Validation.
Refer https://pkg.go.dev/github.com/go-playground/validator/v10#section-readme

File: src/app/authentication/resources/application.yml [hunk 5]
```diff
@@ -2,6 +2,9 @@
 database:
   host: localhost
   port: 3306
-  username: root
+  Username: root
```
Comment (vipin-raj-coditas) [inline src/app/authentication/resources/application.yml]: Keep Username as small case.

**Review by vipin-raj-coditas — CHANGES_REQUESTED** [review]

File: constants/error.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
+package constants
+
+import "errors"
+
+// Custom error types used throughout the application
+var (
+	// ErrUserExistenceFailed indicates a failure in checking user existence.
```
Comment (Gauravi-S) [inline constants/error.go:7]: No need to add comments for every line, if the understanding of code line is obvious.

File: constants/routes.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package constants
+
+const (
+	// User related routes
+	CreateUserRoute = "/signup"
+	SignInRoute     = "/signin"
+)
```
Comment (Gauravi-S) [inline constants/routes.go:7]:
```suggestion
// User related routes (keep the comment here)
const (
	CreateUserRoute = "/signup"
	SignInRoute     = "/signin"
)
```

File: repo/signUpRepo.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+// @return *UserRepository
+func NewUserRepository() *UserRepository {
+	return &UserRepository{} // Return a pointer to a new instance of UserRepository
+}
+
+// CheckIfUserExistsByEmail checks if a user with the given email exists in the database.
+// @param email path string true "User email"
+// @return bool
+// @return error
+func (r *UserRepository) CheckIfUserExistsByEmail(email string) (bool, error) {
```
Comment (Gauravi-S) [inline repo/signUpRepo.go:21]: rename r to appropriate variable name.

File: handlers/signUpHandler.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+// SuccessResponse represents a successful response format.
+type SuccessResponses struct {
+	Message string `json:"message"`
+}
+
+// ErrorResponse represents an error response format.
+type ErrorResponses struct {
+	Error string `json:"error"`
+}
```
Comment (Gauravi-S) [inline handlers/signUpHandler.go:54]: Move both the models SuccessResponses and ErrorResponses in new file genericModels.go in models folder.
same for these models in signInHandler.go

File: models/signUpModel.go [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+// DatabaseConfig represents the configuration values for database connection.
+type DatabaseConfig struct {
+	Host     string `yaml:"host"`
+	Port     int    `yaml:"port"`
+	Username string `yaml:"username"`
+	Password string `yaml:"password"`
+	Name     string `yaml:"name"`
+}
```
Comment (Gauravi-S) [inline models/signUpModel.go:24]: Move this model in new file dbModels.go in models folder.

File: repo/signInRepo.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+// VerifySignInCredentials verifies the user's sign-in credentials against the database.
+// @Summary Verify user sign-in credentials against the database
+// @Description Verifies the user's sign-in credentials against the database.
+// @Accept json
+// @Produce json
+// @Param username path string true "Username"
+// @Param password path string true "Password"
+// @Success 200 {string} string "true"
+// @Failure 400 {string} string "false"
+// @Router /verify [post]
```
Comment (Gauravi-S) [inline repo/signInRepo.go:29]: no need to  add swagger syntax documentation for all the functions, just add one liner description if needed.

File: router/router.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+// @BasePath /
+func SetUpRouter() *gin.Engine {
+	router := gin.Default()
+
+	// Define route for handling POST requests to create new users
+	// @Summary Create a new user
+	// @Description Handles HTTP POST request to create new users
+	// @Accept json
+	// @Produce json
+	// @Router /users [post]
```
Comment (Gauravi-S) [inline router/router.go:28]: remove the documentation.

File: router/router.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	router.POST(constants.CreateUserRoute, handlers.PostUsersData)
```
Comment (Gauravi-S) [inline router/router.go:29]: keep all the router.POST together in code block like

router.POST(constants.CreateUserRoute, handlers.PostUsersData)
router.POST(constants.SignInRoute, handlers.SignInHandler(signInService))

**Pull request closed without merging**